File: 101 isSymmetric/101.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSymmetric(self, root: Optional[TreeNode]) -> bool:
        if not root:
            return True
        stk = []
        stk.append(root.left)
        stk.append(root.right)
        while stk:
            leftNode = stk.pop()
            rightNode = stk.pop()
            if (not leftNode) and (not rightNode):
                continue
            if (not leftNode) or (not rightNode) or (leftNode.val!=rightNode.val):
                return False
            stk.append(leftNode.left)
            stk.append(rightNode.right)
            stk.append(leftNode.right)
            stk.append(rightNode.left)
        return True

File: 101 isSymmetric/101.cpp
//递归方法
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isSymmetric(TreeNode* root) {
        if(root==nullptr)
            return true;
        return compare(root->left,root->right);
    }
    bool compare(TreeNode* left,TreeNode* right){
        if(left==NULL && right==NULL)return true;
        if(left==NULL && right!=NULL)return false;
        if(left!=NULL && right==NULL)return false;
        if(left->val !=right->val)return false;
        bool outside = compare(left->left,right->right);
        bool inside = compare(left->right,right->left);
        return outside&&inside;
    }
};